platform/engineering/design-docs/2021-06-11-small-batch-publishing.md

```diff
@@ -0,0 +1,361 @@
+
+### **Background**
+
+The current content-only build for va.gov runs 9am-12pm Hourly, 1:45pm, 4pm, 5pm ET M-F and generates all HTML pages. The current build takes approximately 15 minutes for page to show up on va.gov. When the amount of content in Drupal reaches 48K nodes, the build will take close to 1 hour to publish content to va.gov.  Editors are often confused about when the content they published in the CMS will be live on va.gov.
+
+Editors coming from Team Sites expect publication of content within 5 minutes.  As editors migrate from Team Sites to Drupal, editors often do not understand why the content they publish does not show up on va.gov in a timely mannter.
+
+
+### **High Level Design**
+
+The same code used to generate the HTML for the preview server can be used to generate the HTML for va.gov.  
+
+The following diagram shows a high level design of how a node in Drupal would be published “immediately” on va.gov.
+
+Drupal is in charge of determining when and which pages need to be published based upon the action by an editor.  When Drupal determines a page is to be published, a Publish Batch Record will be sent to Amazon Simple Queue Service (SQS). A Lambda job will pick up the Publish Batch Record to build the HTML. After the HTML has been created, it’s sent to an [S3 bucket](http://content.www.va.gov.s3-website-us-gov-west-1.amazonaws.com/) which serves the html for va.gov.  
+
+The code used by the lambda to generate the HTML will utilize the current preview server code.  As part of this work, the preview server code will be abstracted and improved to provide production ready HTML.
+
+![image](https://lucid.app/publicSegments/view/84e99c14-7cdf-4955-95b7-badf320be509/image.png)
+
+
+Editable link: [https://lucid.app/lucidchart/a8445048-94ce-4473-aafa-e8ed6cbfbad5/edit?page=0_0#](https://lucid.app/lucidchart/a8445048-94ce-4473-aafa-e8ed6cbfbad5/edit?page=0_0#)
+
+
+## **Specifics**
+
+
+### **Detailed Design**
+
+
+#### Drupal Triggers Content Publishing
+
+When an editor performs an action in Drupal (ie publishing a node, edits a menu item), Drupal will determine which pages need to be published. [The logic will be encapsulated by a Drupal Service.](https://app.zenhub.com/workspaces/vagov-cms-team-5c0e7b864b5806bc2bfc2087/issues/department-of-veterans-affairs/va.gov-cms/5509) The rules around the dependency graph are being tracked on [here](https://github.com/department-of-veterans-affairs/va.gov-team/tree/master/platform/cms/accelerated_publishing/dependency_graph) and are currently a work in progress. Which rules are addressed will be determined by the content type being worked on to support Small Batch Publishing.  As a content type is add to Small Batch Publishing, it will be removed from the hourly content build.  Having support in only one build at a time will elimnte the risk of the full content build overriding the html from the Small Batch Publishing.
```

Comment:
```suggestion
When an editor performs an action in Drupal (ie publishing a node, edits a menu item), Drupal will determine which pages need to be published. [The logic will be encapsulated by a Drupal Service.](https://app.zenhub.com/workspaces/vagov-cms-team-5c0e7b864b5806bc2bfc2087/issues/department-of-veterans-affairs/va.gov-cms/5509) The rules around the dependency graph are being tracked on [here](https://github.com/department-of-veterans-affairs/va.gov-team/tree/master/platform/cms/accelerated_publishing/dependency_graph) and are currently a work in progress. Which rules are addressed will be determined by the content type being worked on to support Small Batch Publishing.  As a content type is add to Small Batch Publishing, it will be removed from the hourly content build.  Having support in only one build at a time will eliminate the risk of the full content build overriding the html from the Small Batch Publishing.
```